### PR TITLE
[shopsys] model service layer removal (phase 4)

### PR DESCRIPTION
**Pull request by DavidOstrozlik**
| Q             | A
| ------------- | ---
|Description, reason for the PR| 
|New feature| No 
|BC breaks| Yes 
|Fixes issues|  
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Model/Order/FrontOrderDataMapper.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use Shopsys\FrameworkBundle\Model\Customer\User;
+
+class FrontOrderDataMapper
+{
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\FrontOrderData $frontOrderData
+     * @param \Shopsys\FrameworkBundle\Model\Customer\User $user
+     * @param \Shopsys\FrameworkBundle\Model\Order\Order $order
+     */
+    public function prefillFrontFormData(FrontOrderData $frontOrderData, User $user, Order $order = null)
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Order/FrontOrderDataMapper.php]: maybe again `?Order $order` instead?

File: docs/upgrade/services_removal.md [hunk 2]
```diff
@@ -121,6 +123,7 @@ Following methods have been removed:
 - `CurrencyService::create()`, use `CurrencyFactory::create()` instead
 - `VatService::getNewDefaultVat()`
 - `OrderProductService::getProductsUsingStockFromOrderProducts()`
+- `OrderCreationService::addOrderItemDiscount()`
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: it was private, therefore I would not mention it here

File: packages/framework/src/Model/Order/Order.php [hunk 3]
```diff
@@ -1095,4 +1097,67 @@ public function fillOrderTransport(
...
+                $quantifiedItemPrice->getUnitPrice(),
+                $product->getVat()->getPercent(),
+                $quantifiedProduct->getQuantity(),
+                $product->getUnit()->getName($locale),
+                $product->getCatnum(),
+                $product
+            );
+
+            if ($quantifiedItemDiscount !== null) {
+                $name = sprintf(
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Order/Order.php]: I would prefer keeping this logic in named protected method

File: packages/framework/src/Model/Order/Order.php [hunk 4]
```diff
@@ -1138,11 +1138,11 @@ public function fillOrderProducts(
 
             if ($quantifiedItemDiscount !== null) {
                 $name = sprintf(
-                   '%s %s - %s',
-                   t('Promo code', [], 'messages', $locale),
-                   $numberFormatterExtension->formatPercent(-$orderPreview->getPromoCodeDiscountPercent(), $locale),
-                   $orderItem->getName()
-               );
+                    '%s %s - %s',
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Order/Order.php]: these changes do not belong to this commit, right? :innocent:

File: docs/upgrade/services_removal.md [hunk 5]
```diff
@@ -131,6 +131,7 @@ Following methods have been removed:
 - `VatService::getNewDefaultVat()`
 - `OrderProductService::getProductsUsingStockFromOrderProducts()`
 - `OrderCreationService::addOrderItemDiscount()`
+- `CartService::clean` use `Cart::clean` instead
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: should be "`CartService::cleanCart()` use `Cart::clean()` instead"

File: project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php [hunk 6]
```diff
@@ -56,4 +56,31 @@ public function testRemoveItem()
         $cart->removeItemById($cartItem1->getId());
         $this->assertSame(1, $cart->getItemsCount());
     }
+
+    public function testCleanCartMakesCartEmpty()
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php]: should be `testCleanMakesCartEmpty()`

File: project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php [hunk 7]
```diff
@@ -56,4 +56,31 @@ public function testRemoveItem()
...
+        $vatData = new VatData();
+        $vatData->name = 'vat';
+        $vatData->percent = 21;
+        $vat = new Vat($vatData);
+
+        $productData = $productDataFactory->create();
+        $productData->name = ['cs' => 'Any name'];
+        $productData->price = $price;
+        $productData->vat = $vat;
+        $product = Product::create($productData);
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php]: Maybe it would be easier to persistence reference instead of creating a product on these 10 lines :slightly_smiling_face:

File: docs/upgrade/services_removal.md [hunk 8]
```diff
@@ -103,6 +103,10 @@ If you use these methods, change their calling appropriately:
   -> `Cart::getCartItemById($cartItemId)`
 - `CartService::getQuantifiedProductsIndexedByCartItemId(Cart $cart)`
   -> `Cart::getQuantifiedProductsIndexedByCartItemId()`
+- `CartService::mergeCarts(Cart $resultingCart, Cart $mergedCart, CustomerIdentifier $customerIdentifier)`
+  -> `Cart::mergeWithCart(Cart $cartToMerge, CartItemFactoryInterface $cartItemFactory, CustomerIdentifier $customerIdentifier)`
+- `CartService::findSimilarCartItemByCartItem(Cart $cart, CartItem $cartItem)`
```
Comment (vitek-rostislav) [inline docs/upgrade/services_removal.md]: was private, no need to mention here

File: project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php [hunk 9]
```diff
@@ -58,6 +60,67 @@ public function testRemoveItem()
...
+        $cartItem = new CartItem($customerIdentifier, $product, 1, '0.0');
+        $cartItems = [$cartItem];
+        $cart = new Cart($cartItems);
+
+        $cart->clean();
+
+        $this->assertTrue($cart->isEmpty());
+    }
+
+    public function testMergeCartsReturnsCartWithSummedProducts()
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php]: should be `testMergeWithCartReturnsCartWithSummedProducts()`

File: packages/framework/src/Model/Cart/Cart.php [hunk 10]
```diff
@@ -151,4 +153,37 @@ protected function findSimilarCartItemByCartItem(CartItem $cartItem)
...
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Customer\CustomerIdentifier $customerIdentifier
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param int $quantity
+     * @param \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPriceCalculationForUser $productPriceCalculation
+     * @param \Shopsys\FrameworkBundle\Model\Cart\Item\CartItemFactoryInterface $cartItemFactory
+     * @return \Shopsys\FrameworkBundle\Model\Cart\AddProductResult
+     */
+    public function addProductToCart(
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Cart/Cart.php]: maybe simply `Cart::addProduct()` would be a better name? :slightly_smiling_face:

File: packages/framework/src/Model/Order/Order.php (hunk 4 above)
Comment (DavidOstrozlik) [reply packages/framework/src/Model/Order/Order.php]: This will be changed in previous commit because it was moved to protected method

File: project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php (hunk 7 above)
Comment (DavidOstrozlik) [reply project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php]: Creating product is changed in next commit to a private method

Comment (vitek-rostislav) [reply project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php]: ...but never mind :laughing:
